### PR TITLE
Add openssl support

### PR DESCRIPTION
**Pull request by YuJuncen**
Hi David, thanks for your awesome library. During using it, we found that the behavior of `rustls` and `openssl` is pretty different, and TiKV uses the `grpcio` gRPC implementation, which wraps `grpc-core` hence uses `BoringSSL` or `openssl` as its transport layer: this made some struggles for users to choose right TLS certificates(some certificates are suitable for `openssl`, some for `rustls`, however now the users can only use the intersection of them). 

So, we have adapted `openssl` via the `hyper-openssl` library, also tried our best to keep the consistency behavior of the `tonic` gRPC implementation.

Generally, we take over the transport layer via the `connect_with_connector` function, however we found that we cannot use custom connector while using `balanced_channel`, so we have manually created a load balancer `OpenSslDiscover`. 

For not changing the old behavior, the openssl support can be enabled via `tls-openssl` or `tls-openssl-vendored`. When either of them enabled, the config `with_openssl_tls` would be enabled, you can then config it via `manually` or provide some cert or keys via some predefined functions.

Perhaps merge this if you think this is helpful. Also, if you have any comments or suggestions about the patch please let me know. =)

### CODE REVIEW

**Comment by davidli2010** [comment]
Thank you for your great job! I will spend some time to review this PR.

**Comment by YuJuncen** [comment]
Hi David, during testing the openssl implementation, we found some problems and updated(The diff is https://github.com/pingcap/etcd-client/pull/3), they are:
- If one of etcd server downs, the load balancer would still choose that node, when it cannot be recovered fast, there might be many failed requests. So we made a `Load` implementation which supports giving up retry over some nodes once detect failure over them.
- During initializing the TLS connection, if there are some errors, the buffered service may fully failed. We need to lazily initializing the connection.

I think the latter should be general-proposed optimization and won't bring any regression, but I'm not sure whether the former is suitable for all scenarios. What do you think? Do we need to adding a feature for disabling it?

**Comment by davidli2010** [comment]
Sorry for my delay, I will take some time at this weekend.

File: Cargo.toml [hunk 1]
```diff
@@ -13,6 +13,8 @@ keywords = ["etcd", "v3", "api", "client", "async"]
 
 [features]
 tls = ["tonic/tls"]
+tls-openssl = ["openssl", "dep:hyper-openssl", "dep:hyper"]
```
Comment (davidli2010) [inline Cargo.toml:16]: Currently, the actual MSRV of `etcd-client` is 1.56. The `dep:` syntax is only available starting with Rust 1.60. So do not use `dep:` here.

File: src/lib.rs [hunk 2]
```diff
@@ -58,12 +58,16 @@
...
+mod channel;
 mod client;
 mod error;
+mod openssl_tls;
 mod rpc;
 
 pub use crate::client::{Client, ConnectOptions};
 pub use crate::error::Error;
+#[cfg(feature = "tls-openssl")]
+pub use crate::openssl_tls::OpenSslClientConfig;
```
Comment (davidli2010) [inline src/lib.rs:70]: The line 70 and 114 can be merged into one line.

File: src/openssl_tls/transport.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#![cfg(feature = "tls-openssl")]
```
Comment (davidli2010) [inline src/openssl_tls/transport.rs:1]: There is `#![cfg(feature = "tls-openssl")]` at the top of `openssl_tls/mod.rs`, is this cfg necessary?

**Review by davidli2010 — CHANGES_REQUESTED** [review]
Frankly I'm not good at openssl, it has been many years since my last coding with openssl. In general this PR is not bad, I think it deserves to merge into master branch.

**Comment by davidli2010** [comment]
> I think the latter should be general-proposed optimization and won't bring any regression, but I'm not sure whether the former is suitable for all scenarios. What do you think? Do we need to adding a feature for disabling it?

I think it's OK as the default behavior. We can add the feature latter if necessary.

**Comment by YuJuncen** [comment]
Sorry but I just found I cannot push to `pingcap/etcd-client` directly, so when I wanna make any change in this PR I must open a PR in `pingcap/etcd-client` firstly, and I found GitHub doesn't support changing the base branch for a opened PR.😢
 
For making easier addressing the comments, I would close this and reopen a PR based on a branch I own. (I will soon address these comments and provide the new PR.)

**Comment by YuJuncen** [comment]
Please review https://github.com/etcdv3/etcd-client/pull/53 (based on a branch I own), sorry for that, I will close this for now.

**Pull request closed without merging**